OpenSim/Tests/Wrapping/testWrapCylinder.cpp

```diff
@@ -0,0 +1,314 @@
+            && IsEqual(lhs.length, rhs.length, tolerance.length)
+            && lhs.noWrap == rhs.noWrap;
+    }
+
+}
+
+// Section on configuring and simulating a specific wrapping scenario.
+namespace {
+
+    // Parameterization of the scenario for testing the wrapping.
+    //
+    // The simulated scene is one with a cylinder at the origin, with a user
+    // defined radius, length, active quadrant, and relative orientation.
+    // A specific wrapping case can be configured by choosing the start and end
+    // points of the path.
+    struct WrapInput final {
+
+        SimTK::Rotation cylinderOrientation() const {
+            SimTK::Rotation orientation;
+            orientation.setRotationToBodyFixedXYZ(eulerRotations);
+            return orientation;
+        }
+
+        // Endpoints of the total path:
+        PathSegment path = {{}, {}};
+
+        // Wrapping cylinder parameters:
+        double radius = 1.;
+        double cylinderLength = 1.;
+        std::string quadrant = "all";
+        SimTK::Vec3 eulerRotations = {0., 0., 0.};
+    };
+
+    std::ostream& operator<<(std::ostream& os, const WrapInput& input) {
+        return os <<
+            "WrapInput{" <<
+            "path: " << input.path << ", " <<
+            "radius: " << input.radius << ", " <<
+            "cylinderLength: " << input.cylinderLength << ", " <<
+            "quadrant: " << input.quadrant << ", " <<
+            "eulerRotations: " << input.eulerRotations << "}";
+    }
+
+    // Simulates the wrapping scenario as configured by the WrapInput.
+    WrapTestResult solve(const WrapInput& input, bool visualize) {
+        Model model;
+        model.setName("testWrapCylinderModel");
+
+        std::unique_ptr<WrapCylinder> cylinder(new WrapCylinder());
+        cylinder->setName("cylinder");
+        cylinder->set_radius(input.radius);
+        cylinder->set_length(input.cylinderLength);
+        cylinder->set_quadrant(input.quadrant);
+        cylinder->set_xyz_body_rotation(input.eulerRotations);
+        cylinder->setFrame(model.getGround());
+
+        std::unique_ptr<PathSpring> spring (new PathSpring("spring", 1., 1., 1.));
+        spring->updGeometryPath().appendNewPathPoint(
+            "startPoint",
+            model.get_ground(),
+            input.path.start);
+        spring->updGeometryPath().appendNewPathPoint(
+            "endPoint",
+            model.get_ground(),
+            input.path.end);
+        spring->updGeometryPath().addPathWrap(*cylinder.get());
+
+        model.addComponent(spring.release());
+        model.updGround().addWrapObject(cylinder.release());
+
+        model.finalizeConnections();
+        model.setUseVisualizer(visualize);
+
+        const SimTK::State& state = model.initSystem();
+
+        if (visualize) {
+            model.realizeVelocity(state);
+            model.getVisualizer().show(state);
+        }
+
+        model.updComponent<PathSpring>("spring").getLength(state);
```

Comment:
Unclear why this is necessary (presumably, you are using it to force some kind of internal caching/realization)?

Reply:
yeah the idea was to trigger the realization. Bit strange that `getLength(state)` is marked as `const` though, not sure if that is common practice?

Reply:
The `const`-ness of a method doesn't apply to `State` realization. It's related to whether or not a method will change the properties of a OpenSim object (e.g., `upd_stiffness` is non-`const` since it modifies the `PathSpring` stiffness).

Reply:
It might be useful to document why it's there or, if this works, explicitly realize the state rather than realizing it via this method (e.g. `model.realizeState(state, SimTK::Stage::Whatever);`)

Reply:
"Realizing" the state will not trigger the computation, hence this sort of manual triggering of the wrap computation. But I added a comment.